server/services/prismic.js

```diff
@@ -259,23 +254,146 @@ function convertPrismicResultsToPaginatedResults(prismicResults: Object): (resul
+  const paginatedResults = convertPrismicResultsToPaginatedResults(events);
   return paginatedResults(promos);
 }
 
-export async function getExhibitionPromos(page: number): Promise<Array<ExhibitionPromo>> {
-  const results = await getAllOfType('exhibitions', page, {orderings: '[my.exhibitions.start]'});
-  const promos = createExhibitionPromos(results);
-  const paginatedResults = convertPrismicResultsToPaginatedResults(results);
+export async function getPaginatedExhibitionPromos(page: number): Promise<Array<ExhibitionPromo>> {
+  const exhibitions = await getAllOfType('exhibitions', page, {orderings: '[my.exhibitions.start]'});
+  const promos = createExhibitionPromos(exhibitions.results);
+  const paginatedResults = convertPrismicResultsToPaginatedResults(exhibitions);
   return paginatedResults(promos);
 }
 
+function datesOverlapRange (eventStartDate, eventEndDate, rangeStartDate, rangeEndDate) {
+  if (rangeStartDate && rangeEndDate) {
+    const eventStart = london(eventStartDate);
+    const eventEnd = london(eventEndDate);
+    const rangeStart = london(rangeStartDate);
+    const rangeEnd = london(rangeEndDate);
+    return (eventStart.isSame(rangeEnd, 'day') || eventStart.isBefore(rangeEnd, 'day')) && (eventEnd.isSame(rangeStart, 'day') || eventEnd.isAfter(rangeStart, 'day'));
+  } else {
+    return true;
+  }
+}
+
+function filterPromosByDate(promos, startDate, endDate) {
+  return promos.filter(e => datesOverlapRange(e.start, e.end, startDate, endDate));
+}
+
+function getActiveState(today, range) {
+  const rangeStart = range && london(range[0]);
+  const rangeEnd = range && london(range[1]);
+  if (rangeStart.isSame(today, 'day') && rangeEnd.isSame(london().add(3, 'month'), 'day')) {
+    return 'everything';
+  } else if (today.isSame(rangeStart, 'day') && today.isSame(rangeEnd, 'day')) {
+    return 'today';
+  } else if (rangeStart.isSame(getWeekendFromDate(today), 'day') && rangeEnd.isSame(getWeekendToDate(today), 'day')) {
+    return 'weekend';
+  }
+};
+
+function groupPromosByMonth(promos) {
+  return promos.reduce((acc, currEvent) => { // TODO tidy this DRY
+    const start = london(currEvent.start);
+    const end = london(currEvent.end);
+    if (end.isSame(start, 'month')) {
+      const year = start.format('YYYY');
+      const month = start.format('MMMM');
+      if (!acc[year]) {
+        acc[year] = {};
+      }
+      if (!acc[year][month]) {
+        acc[year][month] = [];
+      }
+      acc[year][month].push(currEvent);
+    } else {
+      while ((end.isAfter(start) || end.isSame(start, 'month')) && start.isBefore(london().add(3, 'month'))) {
+        const year = start.format('YYYY');
+        const month = start.format('MMMM');
+        if (!acc[year]) {
+          acc[year] = {};
+        }
+        if (!acc[year][month]) {
+          acc[year][month] = [];
+        }
+        acc[year][month].push(currEvent);
+        start.add(1, 'month');
+      }
+    }
+    return acc;
+  }, {});
```

Comment:
A common way to do this is [`groupBy`](http://facebook.github.io/immutable-js/docs/#/List/groupBy)

Reply:
This function isn't actually grouping the promos as such, but duplicating them (if their start and end dates span multiple months/years) and then grouping them accordingly (but the year and month doesn't exist as properties of the events). I've renamed the function so it's clearer.

Reply:
`spanPromoAcrossMonthsAndYears`?
It's still not a concept I'm grabbing intrinsically, but I think I am being dim.

Reply:
Would a diagram similar to what you drew help (it's how I respond well to Rx stuff).
```
                [weekend]
[event1_____]     [event2_____]
``` 

Don't feel obliged, I can just stare at it longer to figure out, I think this is my issue.

Reply:
the diagram was for the datesOverlapRange function and yep I'll add that